// @ng-dynamic-forms/ui-bootstrap/esm5/src/select/dynamic-bootstrap-select.component.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output } from "@angular/core";
import { FormGroup } from "@angular/forms";
import { DynamicFormControlComponent, DynamicFormLayoutService, DynamicFormValidationService, DynamicSelectModel } from "@ng-dynamic-forms/core";
var DynamicBootstrapSelectComponent = /** @class */ (function (_super) {
    __extends(DynamicBootstrapSelectComponent, _super);
    function DynamicBootstrapSelectComponent(layoutService, validationService) {
        var _this = _super.call(this, layoutService, validationService) || this;
        _this.layoutService = layoutService;
        _this.validationService = validationService;
        _this.blur = new EventEmitter();
        _this.change = new EventEmitter();
        _this.focus = new EventEmitter();
        return _this;
    }
    DynamicBootstrapSelectComponent.decorators = [
        { type: Component, args: [{
                    selector: "dynamic-bootstrap-select",
                    template: "<ng-container [formGroup]=\"group\">\n\n    <select class=\"form-control\"\n            [compareWith]=\"model.compareWithFn\"\n            [formControlName]=\"model.id\"\n            [id]=\"elementId\"\n            [name]=\"model.name\"\n            [ngClass]=\"getClass('element', 'control')\"\n            [required]=\"model.required\"\n            [tabindex]=\"model.tabIndex\"\n            (blur)=\"onBlur($event)\"\n            (change)=\"onChange($event)\"\n            (focus)=\"onFocus($event)\">\n\n        <option *ngFor=\"let option of model.options$ | async\"\n                [disabled]=\"option.disabled\"\n                [ngValue]=\"option.value\">{{ option.label }}\n        </option>\n\n    </select>\n\n</ng-container>\n",
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    /** @nocollapse */
    DynamicBootstrapSelectComponent.ctorParameters = function () { return [
        { type: DynamicFormLayoutService },
        { type: DynamicFormValidationService }
    ]; };
    DynamicBootstrapSelectComponent.propDecorators = {
        group: [{ type: Input }],
        layout: [{ type: Input }],
        model: [{ type: Input }],
        blur: [{ type: Output }],
        change: [{ type: Output }],
        focus: [{ type: Output }]
    };
    return DynamicBootstrapSelectComponent;
}(DynamicFormControlComponent));
export { DynamicBootstrapSelectComponent };
if (false) {
    /** @type {?} */
    DynamicBootstrapSelectComponent.prototype.group;
    /** @type {?} */
    DynamicBootstrapSelectComponent.prototype.layout;
    /** @type {?} */
    DynamicBootstrapSelectComponent.prototype.model;
    /** @type {?} */
    DynamicBootstrapSelectComponent.prototype.blur;
    /** @type {?} */
    DynamicBootstrapSelectComponent.prototype.change;
    /** @type {?} */
    DynamicBootstrapSelectComponent.prototype.focus;
    /**
     * @type {?}
     * @protected
     */
    DynamicBootstrapSelectComponent.prototype.layoutService;
    /**
     * @type {?}
     * @protected
     */
    DynamicBootstrapSelectComponent.prototype.validationService;
}
