// @ng-dynamic-forms/ui-bootstrap/esm5/src/rating/dynamic-bootstrap-rating.component.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output } from "@angular/core";
import { FormGroup } from "@angular/forms";
import { DynamicFormLayoutService, DynamicFormValidationService, DynamicFormControlComponent, DynamicRatingModel } from "@ng-dynamic-forms/core";
var DynamicBootstrapRatingComponent = /** @class */ (function (_super) {
    __extends(DynamicBootstrapRatingComponent, _super);
    function DynamicBootstrapRatingComponent(layoutService, validationService) {
        var _this = _super.call(this, layoutService, validationService) || this;
        _this.layoutService = layoutService;
        _this.validationService = validationService;
        _this.blur = new EventEmitter();
        _this.change = new EventEmitter();
        _this.focus = new EventEmitter();
        return _this;
    }
    DynamicBootstrapRatingComponent.decorators = [
        { type: Component, args: [{
                    selector: "dynamic-bootstrap-rating",
                    template: "<ng-container [formGroup]=\"group\">\n\n    <rating [formControlName]=\"model.id\"\n            [id]=\"elementId\"\n            [max]=\"model.max\"\n            [ngClass]=\"getClass('element', 'control')\"\n            [readonly]=\"model.getAdditional('readonly', false)\"\n            (onHover)=\"onCustomEvent($event, 'onHover')\"\n            (onLeave)=\"onCustomEvent($event, 'onLeave')\"></rating>\n\n</ng-container>\n",
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    /** @nocollapse */
    DynamicBootstrapRatingComponent.ctorParameters = function () { return [
        { type: DynamicFormLayoutService },
        { type: DynamicFormValidationService }
    ]; };
    DynamicBootstrapRatingComponent.propDecorators = {
        group: [{ type: Input }],
        layout: [{ type: Input }],
        model: [{ type: Input }],
        blur: [{ type: Output }],
        change: [{ type: Output }],
        focus: [{ type: Output }]
    };
    return DynamicBootstrapRatingComponent;
}(DynamicFormControlComponent));
export { DynamicBootstrapRatingComponent };
if (false) {
    /** @type {?} */
    DynamicBootstrapRatingComponent.prototype.group;
    /** @type {?} */
    DynamicBootstrapRatingComponent.prototype.layout;
    /** @type {?} */
    DynamicBootstrapRatingComponent.prototype.model;
    /** @type {?} */
    DynamicBootstrapRatingComponent.prototype.blur;
    /** @type {?} */
    DynamicBootstrapRatingComponent.prototype.change;
    /** @type {?} */
    DynamicBootstrapRatingComponent.prototype.focus;
    /**
     * @type {?}
     * @protected
     */
    DynamicBootstrapRatingComponent.prototype.layoutService;
    /**
     * @type {?}
     * @protected
     */
    DynamicBootstrapRatingComponent.prototype.validationService;
}
