// @ng-dynamic-forms/ui-bootstrap/esm5/src/dynamic-bootstrap-form-control-container.component.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ComponentFactoryResolver, ContentChildren, EventEmitter, Input, Output, ViewChild, ViewContainerRef } from "@angular/core";
import { FormGroup } from "@angular/forms";
import { DYNAMIC_FORM_CONTROL_TYPE_ARRAY, DYNAMIC_FORM_CONTROL_TYPE_CHECKBOX, DYNAMIC_FORM_CONTROL_TYPE_CHECKBOX_GROUP, DYNAMIC_FORM_CONTROL_TYPE_DATEPICKER, DYNAMIC_FORM_CONTROL_TYPE_GROUP, DYNAMIC_FORM_CONTROL_TYPE_INPUT, DYNAMIC_FORM_CONTROL_TYPE_RADIO_GROUP, DYNAMIC_FORM_CONTROL_TYPE_RATING, DYNAMIC_FORM_CONTROL_TYPE_SELECT, DYNAMIC_FORM_CONTROL_TYPE_TEXTAREA, DYNAMIC_FORM_CONTROL_TYPE_TIMEPICKER, DynamicFormControlContainerComponent, DynamicFormControlModel, DynamicFormInstancesService, DynamicFormLayoutService, DynamicFormValidationService, DynamicTemplateDirective } from "@ng-dynamic-forms/core";
import { DynamicBootstrapCheckboxComponent } from "./checkbox/dynamic-bootstrap-checkbox.component";
import { DynamicBootstrapDatePickerComponent } from "./datepicker/dynamic-bootstrap-datepicker.component";
import { DynamicBootstrapFormArrayComponent } from "./form-array/dynamic-bootstrap-form-array.component";
import { DynamicBootstrapFormGroupComponent } from "./form-group/dynamic-bootstrap-form-group.component";
import { DynamicBootstrapInputComponent } from "./input/dynamic-bootstrap-input.component";
import { DynamicBootstrapRadioGroupComponent } from "./radio-group/dynamic-bootstrap-radio-group.component";
import { DynamicBootstrapRatingComponent } from "./rating/dynamic-bootstrap-rating.component";
import { DynamicBootstrapSelectComponent } from "./select/dynamic-bootstrap-select.component";
import { DynamicBootstrapTextAreaComponent } from "./textarea/dynamic-bootstrap-textarea.component";
import { DynamicBootstrapTimePickerComponent } from "./timepicker/dynamic-bootstrap-timepicker.component";
var DynamicBootstrapFormControlContainerComponent = /** @class */ (function (_super) {
    __extends(DynamicBootstrapFormControlContainerComponent, _super);
    function DynamicBootstrapFormControlContainerComponent(componentFactoryResolver, layoutService, validationService, dynamicFormInstancesService) {
        var _this = _super.call(this, componentFactoryResolver, layoutService, validationService, dynamicFormInstancesService) || this;
        _this.componentFactoryResolver = componentFactoryResolver;
        _this.layoutService = layoutService;
        _this.validationService = validationService;
        _this.dynamicFormInstancesService = dynamicFormInstancesService;
        _this.asBootstrapFormGroup = true;
        _this.context = null;
        _this.blur = new EventEmitter();
        _this.change = new EventEmitter();
        _this.focus = new EventEmitter();
        _this.customEvent = new EventEmitter();
        return _this;
    }
    Object.defineProperty(DynamicBootstrapFormControlContainerComponent.prototype, "componentType", {
        get: /**
         * @return {?}
         */
        function () {
            return this.layoutService.getCustomComponentType(this.model) || bootstrapUIFormControlMapFn(this.model);
        },
        enumerable: true,
        configurable: true
    });
    DynamicBootstrapFormControlContainerComponent.decorators = [
        { type: Component, args: [{
                    selector: "dynamic-bootstrap-form-control",
                    template: "<div [class.form-group]=\"asBootstrapFormGroup || getClass('element', 'container').includes('form-group')\"\n     [class.has-error]=\"showErrorMessages\"\n     [formGroup]=\"group\"\n     [ngClass]=\"[getClass('element', 'container'), getClass('grid', 'container')]\">\n\n    <label *ngIf=\"!isCheckbox && hasLabel\"\n           [for]=\"elementId\"\n           [innerHTML]=\"model.label\"\n           [ngClass]=\"[getClass('element', 'label'), getClass('grid', 'label')]\"></label>\n\n    <ng-container *ngTemplateOutlet=\"startTemplate?.templateRef; context: model\"></ng-container>\n\n    <div [ngClass]=\"getClass('grid', 'control')\">\n\n        <ng-container #componentViewContainer></ng-container>\n\n        <span *ngIf=\"hasHint\" class=\"help-block\" [innerHTML]=\"hint\"\n              [ngClass]=\"getClass('element', 'hint')\"></span>\n\n    </div>\n\n    <ul *ngIf=\"showErrorMessages\" class=\"has-error\"\n        [ngClass]=\"[getClass('element', 'errors'), getClass('grid', 'errors')]\">\n\n        <li *ngFor=\"let message of errorMessages\" class=\"help-block\">{{ message }}</li>\n\n    </ul>\n\n    <ng-container *ngTemplateOutlet=\"endTemplate?.templateRef; context: model\"></ng-container>\n\n    <ng-content></ng-content>\n\n</div>\n"
                }] }
    ];
    /** @nocollapse */
    DynamicBootstrapFormControlContainerComponent.ctorParameters = function () { return [
        { type: ComponentFactoryResolver },
        { type: DynamicFormLayoutService },
        { type: DynamicFormValidationService },
        { type: DynamicFormInstancesService }
    ]; };
    DynamicBootstrapFormControlContainerComponent.propDecorators = {
        contentTemplateList: [{ type: ContentChildren, args: [DynamicTemplateDirective,] }],
        inputTemplateList: [{ type: Input, args: ["templates",] }],
        asBootstrapFormGroup: [{ type: Input }],
        context: [{ type: Input }],
        group: [{ type: Input }],
        layout: [{ type: Input }],
        model: [{ type: Input }],
        blur: [{ type: Output }],
        change: [{ type: Output }],
        focus: [{ type: Output }],
        customEvent: [{ type: Output, args: ["bsEvent",] }],
        componentViewContainerRef: [{ type: ViewChild, args: ["componentViewContainer", { read: ViewContainerRef },] }]
    };
    return DynamicBootstrapFormControlContainerComponent;
}(DynamicFormControlContainerComponent));
export { DynamicBootstrapFormControlContainerComponent };
if (false) {
    /** @type {?} */
    DynamicBootstrapFormControlContainerComponent.prototype.contentTemplateList;
    /** @type {?} */
    DynamicBootstrapFormControlContainerComponent.prototype.inputTemplateList;
    /** @type {?} */
    DynamicBootstrapFormControlContainerComponent.prototype.asBootstrapFormGroup;
    /** @type {?} */
    DynamicBootstrapFormControlContainerComponent.prototype.context;
    /** @type {?} */
    DynamicBootstrapFormControlContainerComponent.prototype.group;
    /** @type {?} */
    DynamicBootstrapFormControlContainerComponent.prototype.layout;
    /** @type {?} */
    DynamicBootstrapFormControlContainerComponent.prototype.model;
    /** @type {?} */
    DynamicBootstrapFormControlContainerComponent.prototype.blur;
    /** @type {?} */
    DynamicBootstrapFormControlContainerComponent.prototype.change;
    /** @type {?} */
    DynamicBootstrapFormControlContainerComponent.prototype.focus;
    /** @type {?} */
    DynamicBootstrapFormControlContainerComponent.prototype.customEvent;
    /** @type {?} */
    DynamicBootstrapFormControlContainerComponent.prototype.componentViewContainerRef;
    /**
     * @type {?}
     * @protected
     */
    DynamicBootstrapFormControlContainerComponent.prototype.componentFactoryResolver;
    /**
     * @type {?}
     * @protected
     */
    DynamicBootstrapFormControlContainerComponent.prototype.layoutService;
    /**
     * @type {?}
     * @protected
     */
    DynamicBootstrapFormControlContainerComponent.prototype.validationService;
    /**
     * @type {?}
     * @protected
     */
    DynamicBootstrapFormControlContainerComponent.prototype.dynamicFormInstancesService;
}
/**
 * @param {?} model
 * @return {?}
 */
export function bootstrapUIFormControlMapFn(model) {
    switch (model.type) {
        case DYNAMIC_FORM_CONTROL_TYPE_ARRAY:
            return DynamicBootstrapFormArrayComponent;
        case DYNAMIC_FORM_CONTROL_TYPE_CHECKBOX:
            return DynamicBootstrapCheckboxComponent;
        case DYNAMIC_FORM_CONTROL_TYPE_CHECKBOX_GROUP:
            return DynamicBootstrapFormGroupComponent;
        case DYNAMIC_FORM_CONTROL_TYPE_DATEPICKER:
            return DynamicBootstrapDatePickerComponent;
        case DYNAMIC_FORM_CONTROL_TYPE_GROUP:
            return DynamicBootstrapFormGroupComponent;
        case DYNAMIC_FORM_CONTROL_TYPE_INPUT:
            return DynamicBootstrapInputComponent;
        case DYNAMIC_FORM_CONTROL_TYPE_RADIO_GROUP:
            return DynamicBootstrapRadioGroupComponent;
        case DYNAMIC_FORM_CONTROL_TYPE_RATING:
            return DynamicBootstrapRatingComponent;
        case DYNAMIC_FORM_CONTROL_TYPE_SELECT:
            return DynamicBootstrapSelectComponent;
        case DYNAMIC_FORM_CONTROL_TYPE_TEXTAREA:
            return DynamicBootstrapTextAreaComponent;
        case DYNAMIC_FORM_CONTROL_TYPE_TIMEPICKER:
            return DynamicBootstrapTimePickerComponent;
        default:
            return null;
    }
}
