// @ng-dynamic-forms/ui-bootstrap/bundles/ui-bootstrap.umd.js
/*!
@ng-dynamic-forms/ui-bootstrap 7.1.1 2019-03-14 19:20 UTC
Copyright (c) 2016-2018, Udo Schöfer http://www.udos86.de

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted, provided that the above
copyright notice and this permission notice appear in all copies.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
*/
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/forms'), require('@ng-dynamic-forms/core'), require('ngx-bootstrap/datepicker'), require('ngx-bootstrap/timepicker'), require('@angular/common'), require('angular2-text-mask')) :
    typeof define === 'function' && define.amd ? define(['exports', '@angular/core', '@angular/forms', '@ng-dynamic-forms/core', 'ngx-bootstrap/datepicker', 'ngx-bootstrap/timepicker', '@angular/common', 'angular2-text-mask'], factory) :
    (global = global || self, factory((global.ngDF = global.ngDF || {}, global.ngDF.bootstrapUI = {}), global.ng.core, global.ng.forms, global.ngDF.core, global['ngx-bootstrap'].umd, global['ngx-bootstrap'].umd, global.ng.common, global['angular2-text-mask']));
}(this, function (exports, core, forms, core$1, datepicker, timepicker, common, angular2TextMask) { 'use strict';

    var __extends = (this && this.__extends) || (function () {
        var extendStatics = function (d, b) {
            extendStatics = Object.setPrototypeOf ||
                ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
                function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
            return extendStatics(d, b);
        };
        return function (d, b) {
            extendStatics(d, b);
            function __() { this.constructor = d; }
            d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
        };
    })();
    var DynamicBootstrapCheckboxComponent = /** @class */ (function (_super) {
        __extends(DynamicBootstrapCheckboxComponent, _super);
        function DynamicBootstrapCheckboxComponent(layoutService, validationService) {
            var _this = _super.call(this, layoutService, validationService) || this;
            _this.layoutService = layoutService;
            _this.validationService = validationService;
            _this.blur = new core.EventEmitter();
            _this.change = new core.EventEmitter();
            _this.focus = new core.EventEmitter();
            return _this;
        }
        DynamicBootstrapCheckboxComponent.decorators = [
            { type: core.Component, args: [{
                        selector: "dynamic-bootstrap-checkbox",
                        template: "<div [formGroup]=\"group\" [class.disabled]=\"model.disabled\" [ngClass]=\"getClass('element', 'control') || 'checkbox'\">\n\n    <label>\n\n        <input type=\"checkbox\"\n               [checked]=\"model.checked\"\n               [formControlName]=\"model.id\"\n               [id]=\"elementId\"\n               [indeterminate]=\"model.indeterminate\"\n               [name]=\"model.name\"\n               [ngClass]=\"getClass('element', 'control')\"\n               [required]=\"model.required\"\n               [tabindex]=\"model.tabIndex\"\n               [value]=\"model.value\"\n               (blur)=\"onBlur($event)\"\n               (change)=\"onChange($event)\"\n               (focus)=\"onFocus($event)\"/><span [innerHTML]=\"model.label\"\n                                                [ngClass]=\"[getClass('element', 'label'), getClass('grid', 'label')]\"></span></label>\n</div>\n",
                        changeDetection: core.ChangeDetectionStrategy.OnPush
                    }] }
        ];
        /** @nocollapse */
        DynamicBootstrapCheckboxComponent.ctorParameters = function () { return [
            { type: core$1.DynamicFormLayoutService },
            { type: core$1.DynamicFormValidationService }
        ]; };
        DynamicBootstrapCheckboxComponent.propDecorators = {
            group: [{ type: core.Input }],
            layout: [{ type: core.Input }],
            model: [{ type: core.Input }],
            blur: [{ type: core.Output }],
            change: [{ type: core.Output }],
            focus: [{ type: core.Output }]
        };
        return DynamicBootstrapCheckboxComponent;
    }(core$1.DynamicFormControlComponent));

    var __extends$1 = (this && this.__extends) || (function () {
        var extendStatics = function (d, b) {
            extendStatics = Object.setPrototypeOf ||
                ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
                function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
            return extendStatics(d, b);
        };
        return function (d, b) {
            extendStatics(d, b);
            function __() { this.constructor = d; }
            d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
        };
    })();
    var DynamicBootstrapDatePickerComponent = /** @class */ (function (_super) {
        __extends$1(DynamicBootstrapDatePickerComponent, _super);
        function DynamicBootstrapDatePickerComponent(layoutService, validationService) {
            var _this = _super.call(this, layoutService, validationService) || this;
            _this.layoutService = layoutService;
            _this.validationService = validationService;
            _this.blur = new core.EventEmitter();
            _this.change = new core.EventEmitter();
            _this.customEvent = new core.EventEmitter();
            _this.focus = new core.EventEmitter();
            return _this;
        }
        DynamicBootstrapDatePickerComponent.decorators = [
            { type: core.Component, args: [{
                        selector: "dynamic-bootstrap-datepicker",
                        template: "<div [formGroup]=\"group\" class=\"input-group\">\n\n    <input bsDatepicker class=\"form-control\" #bsDatepicker=\"bsDatepicker\"\n           [autofocus]=\"model.autoFocus\"\n           [bsConfig]=\"model.getAdditional('bsConfig', {containerClass: model.getAdditional('containerClass', 'theme-green'), showWeekNumbers: model.getAdditional('showWeekNumbers', true)})\"\n           [formControlName]=\"model.id\"\n           [id]=\"elementId\"\n           [maxDate]=\"model.max\"\n           [minDate]=\"model.min\"\n           [name]=\"model.name\"\n           [ngClass]=\"getClass('element', 'control')\"\n           [outsideClick]=\"model.getAdditional('outsideClick', true)\"\n           [placeholder]=\"model.placeholder\"\n           [placement]=\"model.getAdditional('placement', 'bottom')\"\n           [readonly]=\"model.readOnly\"\n           [required]=\"model.required\"\n           (bsValueChange)=\"onChange($event)\"\n           (onHidden)=\"onCustomEvent($event, 'onHidden')\"\n           (onShown)=\"onCustomEvent($event, 'onShown')\">\n\n    <div *ngIf=\"model.toggleLabel\" class=\"input-group-btn\">\n\n        <button *ngIf=\"model.toggleIcon\" class=\"btn\" (click)=\"bsDatepicker.toggle()\"><img [src]=\"model.toggleIcon\"/></button>\n        <button *ngIf=\"model.toggleLabel\" class=\"btn\" (click)=\"bsDatepicker.toggle()\">{{ model.toggleLabel }}</button>\n\n    </div>\n\n</div>\n",
                        changeDetection: core.ChangeDetectionStrategy.OnPush
                    }] }
        ];
        /** @nocollapse */
        DynamicBootstrapDatePickerComponent.ctorParameters = function () { return [
            { type: core$1.DynamicFormLayoutService },
            { type: core$1.DynamicFormValidationService }
        ]; };
        DynamicBootstrapDatePickerComponent.propDecorators = {
            group: [{ type: core.Input }],
            layout: [{ type: core.Input }],
            model: [{ type: core.Input }],
            blur: [{ type: core.Output }],
            change: [{ type: core.Output }],
            customEvent: [{ type: core.Output }],
            focus: [{ type: core.Output }],
            bsDatePicker: [{ type: core.ViewChild, args: [datepicker.BsDatepickerDirective,] }]
        };
        return DynamicBootstrapDatePickerComponent;
    }(core$1.DynamicFormControlComponent));

    var __extends$2 = (this && this.__extends) || (function () {
        var extendStatics = function (d, b) {
            extendStatics = Object.setPrototypeOf ||
                ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
                function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
            return extendStatics(d, b);
        };
        return function (d, b) {
            extendStatics(d, b);
            function __() { this.constructor = d; }
            d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
        };
    })();
    var DynamicBootstrapFormArrayComponent = /** @class */ (function (_super) {
        __extends$2(DynamicBootstrapFormArrayComponent, _super);
        function DynamicBootstrapFormArrayComponent(layoutService, validationService) {
            var _this = _super.call(this, layoutService, validationService) || this;
            _this.layoutService = layoutService;
            _this.validationService = validationService;
            _this.blur = new core.EventEmitter();
            _this.change = new core.EventEmitter();
            _this.customEvent = new core.EventEmitter();
            _this.focus = new core.EventEmitter();
            return _this;
        }
        DynamicBootstrapFormArrayComponent.decorators = [
            { type: core.Component, args: [{
                        selector: "dynamic-bootstrap-form-array",
                        template: "<ng-container [formGroup]=\"group\">\n\n    <div [formArrayName]=\"model.id\" [id]=\"elementId\" [ngClass]=\"getClass('element', 'control')\">\n\n        <div *ngFor=\"let groupModel of model.groups; let idx = index\" role=\"group\"\n             [formGroupName]=\"idx\" [ngClass]=\"[getClass('element', 'group'), getClass('grid', 'group')]\">\n\n            <ng-container *ngTemplateOutlet=\"startTemplate?.templateRef; context: groupModel\"></ng-container>\n\n            <dynamic-bootstrap-form-control *ngFor=\"let _model of groupModel.group\"\n                                            [context]=\"groupModel\"\n                                            [group]=\"array.at(idx)\"\n                                            [hidden]=\"_model.hidden\"\n                                            [layout]=\"layout\"\n                                            [model]=\"_model\"\n                                            [templates]=\"templates\"\n                                            [ngClass]=\"[getClass('element', 'host', _model), getClass('grid', 'host', _model),\n                                                        getClass('element','children'), getClass('grid','children')]\"\n                                            (blur)=\"onBlur($event)\"\n                                            (change)=\"onChange($event)\"\n                                            (focus)=\"onFocus($event)\"\n                                            (bsEvent)=\"onCustomEvent($event, null, true)\"></dynamic-bootstrap-form-control>\n\n            <ng-container *ngTemplateOutlet=\"endTemplate?.templateRef; context: groupModel\"></ng-container>\n\n        </div>\n\n    </div>\n\n</ng-container>\n"
                    }] }
        ];
        /** @nocollapse */
        DynamicBootstrapFormArrayComponent.ctorParameters = function () { return [
            { type: core$1.DynamicFormLayoutService },
            { type: core$1.DynamicFormValidationService }
        ]; };
        DynamicBootstrapFormArrayComponent.propDecorators = {
            group: [{ type: core.Input }],
            layout: [{ type: core.Input }],
            model: [{ type: core.Input }],
            templates: [{ type: core.Input }],
            blur: [{ type: core.Output }],
            change: [{ type: core.Output }],
            customEvent: [{ type: core.Output }],
            focus: [{ type: core.Output }]
        };
        return DynamicBootstrapFormArrayComponent;
    }(core$1.DynamicFormArrayComponent));

    var __extends$3 = (this && this.__extends) || (function () {
        var extendStatics = function (d, b) {
            extendStatics = Object.setPrototypeOf ||
                ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
                function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
            return extendStatics(d, b);
        };
        return function (d, b) {
            extendStatics(d, b);
            function __() { this.constructor = d; }
            d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
        };
    })();
    var DynamicBootstrapFormGroupComponent = /** @class */ (function (_super) {
        __extends$3(DynamicBootstrapFormGroupComponent, _super);
        function DynamicBootstrapFormGroupComponent(layoutService, validationService) {
            var _this = _super.call(this, layoutService, validationService) || this;
            _this.layoutService = layoutService;
            _this.validationService = validationService;
            _this.blur = new core.EventEmitter();
            _this.change = new core.EventEmitter();
            _this.customEvent = new core.EventEmitter();
            _this.focus = new core.EventEmitter();
            return _this;
        }
        DynamicBootstrapFormGroupComponent.decorators = [
            { type: core.Component, args: [{
                        selector: "dynamic-bootstrap-form-group",
                        template: "<ng-container [formGroup]=\"group\">\n\n    <fieldset [formGroupName]=\"model.id\"\n              [id]=\"elementId\"\n              [name]=\"model.name\"\n              [ngClass]=\"getClass('element', 'control')\">\n\n        <legend *ngIf=\"model.legend\" [innerHTML]=\"model.legend\"></legend>\n\n        <dynamic-bootstrap-form-control *ngFor=\"let _model of model.group\"\n                                        [asBootstrapFormGroup]=\"false\"\n                                        [group]=\"control\"\n                                        [hidden]=\"_model.hidden\"\n                                        [layout]=\"layout\"\n                                        [model]=\"_model\"\n                                        [templates]=\"templates\"\n                                        [ngClass]=\"[getClass('element', 'host', _model), getClass('grid', 'host', _model),\n                                                    getClass('element','children'), getClass('grid','children')]\"\n                                        (blur)=\"onBlur($event)\"\n                                        (change)=\"onChange($event)\"\n                                        (focus)=\"onFocus($event)\"\n                                        (bsEvent)=\"onCustomEvent($event, null, true)\"></dynamic-bootstrap-form-control>\n    </fieldset>\n\n</ng-container>\n"
                    }] }
        ];
        /** @nocollapse */
        DynamicBootstrapFormGroupComponent.ctorParameters = function () { return [
            { type: core$1.DynamicFormLayoutService },
            { type: core$1.DynamicFormValidationService }
        ]; };
        DynamicBootstrapFormGroupComponent.propDecorators = {
            group: [{ type: core.Input }],
            layout: [{ type: core.Input }],
            model: [{ type: core.Input }],
            templates: [{ type: core.Input }],
            blur: [{ type: core.Output }],
            change: [{ type: core.Output }],
            customEvent: [{ type: core.Output }],
            focus: [{ type: core.Output }]
        };
        return DynamicBootstrapFormGroupComponent;
    }(core$1.DynamicFormControlComponent));

    var __extends$4 = (this && this.__extends) || (function () {
        var extendStatics = function (d, b) {
            extendStatics = Object.setPrototypeOf ||
                ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
                function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
            return extendStatics(d, b);
        };
        return function (d, b) {
            extendStatics(d, b);
            function __() { this.constructor = d; }
            d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
        };
    })();
    var DynamicBootstrapInputComponent = /** @class */ (function (_super) {
        __extends$4(DynamicBootstrapInputComponent, _super);
        function DynamicBootstrapInputComponent(layoutService, validationService) {
            var _this = _super.call(this, layoutService, validationService) || this;
            _this.layoutService = layoutService;
            _this.validationService = validationService;
            _this.blur = new core.EventEmitter();
            _this.change = new core.EventEmitter();
            _this.focus = new core.EventEmitter();
            return _this;
        }
        DynamicBootstrapInputComponent.decorators = [
            { type: core.Component, args: [{
                        selector: "dynamic-bootstrap-input",
                        template: "<div [formGroup]=\"group\" [class.input-group]=\"model.prefix || model.suffix\">\n\n    <div *ngIf=\"model.prefix\" class=\"input-group-addon\" [innerHTML]=\"model.prefix\"></div>\n\n    <input class=\"form-control\"\n           [attr.accept]=\"model.accept\"\n           [attr.max]=\"model.max\"\n           [attr.min]=\"model.min\"\n           [attr.multiple]=\"model.multiple\"\n           [attr.step]=\"model.step\"\n           [autocomplete]=\"model.autoComplete\"\n           [autofocus]=\"model.autoFocus\"\n           [dynamicList]=\"model.listId\"\n           [formControlName]=\"model.id\"\n           [id]=\"elementId\"\n           [maxlength]=\"model.maxLength\"\n           [minlength]=\"model.minLength\"\n           [name]=\"model.name\"\n           [ngClass]=\"getClass('element', 'control')\"\n           [pattern]=\"model.pattern\"\n           [placeholder]=\"model.placeholder\"\n           [readonly]=\"model.readOnly\"\n           [required]=\"model.required\"\n           [spellcheck]=\"model.spellCheck\"\n           [tabindex]=\"model.tabIndex\"\n           [textMask]=\"{mask: (model.mask || false), showMask: model.mask && !(model.placeholder)}\"\n           [type]=\"model.inputType\"\n           (blur)=\"onBlur($event)\"\n           (change)=\"onChange($event)\"\n           (focus)=\"onFocus($event)\"/>\n\n    <div *ngIf=\"model.suffix\" class=\"input-group-addon\" [innerHTML]=\"model.suffix\"></div>\n\n    <datalist *ngIf=\"model.hasList\" [id]=\"model.listId\">\n\n        <option *ngFor=\"let option of model.list$ | async\" [value]=\"option\">\n\n    </datalist>\n\n</div>\n",
                        changeDetection: core.ChangeDetectionStrategy.OnPush
                    }] }
        ];
        /** @nocollapse */
        DynamicBootstrapInputComponent.ctorParameters = function () { return [
            { type: core$1.DynamicFormLayoutService },
            { type: core$1.DynamicFormValidationService }
        ]; };
        DynamicBootstrapInputComponent.propDecorators = {
            group: [{ type: core.Input }],
            layout: [{ type: core.Input }],
            model: [{ type: core.Input }],
            blur: [{ type: core.Output }],
            change: [{ type: core.Output }],
            focus: [{ type: core.Output }]
        };
        return DynamicBootstrapInputComponent;
    }(core$1.DynamicFormControlComponent));

    var __extends$5 = (this && this.__extends) || (function () {
        var extendStatics = function (d, b) {
            extendStatics = Object.setPrototypeOf ||
                ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
                function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
            return extendStatics(d, b);
        };
        return function (d, b) {
            extendStatics(d, b);
            function __() { this.constructor = d; }
            d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
        };
    })();
    var DynamicBootstrapRadioGroupComponent = /** @class */ (function (_super) {
        __extends$5(DynamicBootstrapRadioGroupComponent, _super);
        function DynamicBootstrapRadioGroupComponent(layoutService, validationService) {
            var _this = _super.call(this, layoutService, validationService) || this;
            _this.layoutService = layoutService;
            _this.validationService = validationService;
            _this.blur = new core.EventEmitter();
            _this.change = new core.EventEmitter();
            _this.focus = new core.EventEmitter();
            return _this;
        }
        DynamicBootstrapRadioGroupComponent.decorators = [
            { type: core.Component, args: [{
                        selector: "dynamic-bootstrap-radio-group",
                        template: "<fieldset role=\"radiogroup\"\n          [formGroup]=\"group\"\n          [id]=\"elementId\"\n          [name]=\"model.name\"\n          [ngClass]=\"getClass('element', 'control')\"\n          [tabindex]=\"model.tabIndex\"\n          (change)=\"onChange($event)\">\n\n    <legend *ngIf=\"model.legend\" [innerHTML]=\"model.legend\"></legend>\n\n    <div *ngFor=\"let option of model.options$ | async\" [ngClass]=\"getClass('element', 'option') || 'radio'\">\n\n        <label><input type=\"radio\"\n                      [formControlName]=\"model.id\"\n                      [name]=\"model.name\"\n                      [value]=\"option.value\"\n                      (blur)=\"onBlur($event)\"\n                      (focus)=\"onFocus($event)\"/><span [innerHTML]=\"option.label\"></span></label>\n    </div>\n\n</fieldset>\n",
                        changeDetection: core.ChangeDetectionStrategy.OnPush
                    }] }
        ];
        /** @nocollapse */
        DynamicBootstrapRadioGroupComponent.ctorParameters = function () { return [
            { type: core$1.DynamicFormLayoutService },
            { type: core$1.DynamicFormValidationService }
        ]; };
        DynamicBootstrapRadioGroupComponent.propDecorators = {
            group: [{ type: core.Input }],
            layout: [{ type: core.Input }],
            model: [{ type: core.Input }],
            blur: [{ type: core.Output }],
            change: [{ type: core.Output }],
            focus: [{ type: core.Output }]
        };
        return DynamicBootstrapRadioGroupComponent;
    }(core$1.DynamicFormControlComponent));

    var __extends$6 = (this && this.__extends) || (function () {
        var extendStatics = function (d, b) {
            extendStatics = Object.setPrototypeOf ||
                ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
                function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
            return extendStatics(d, b);
        };
        return function (d, b) {
            extendStatics(d, b);
            function __() { this.constructor = d; }
            d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
        };
    })();
    var DynamicBootstrapRatingComponent = /** @class */ (function (_super) {
        __extends$6(DynamicBootstrapRatingComponent, _super);
        function DynamicBootstrapRatingComponent(layoutService, validationService) {
            var _this = _super.call(this, layoutService, validationService) || this;
            _this.layoutService = layoutService;
            _this.validationService = validationService;
            _this.blur = new core.EventEmitter();
            _this.change = new core.EventEmitter();
            _this.focus = new core.EventEmitter();
            return _this;
        }
        DynamicBootstrapRatingComponent.decorators = [
            { type: core.Component, args: [{
                        selector: "dynamic-bootstrap-rating",
                        template: "<ng-container [formGroup]=\"group\">\n\n    <rating [formControlName]=\"model.id\"\n            [id]=\"elementId\"\n            [max]=\"model.max\"\n            [ngClass]=\"getClass('element', 'control')\"\n            [readonly]=\"model.getAdditional('readonly', false)\"\n            (onHover)=\"onCustomEvent($event, 'onHover')\"\n            (onLeave)=\"onCustomEvent($event, 'onLeave')\"></rating>\n\n</ng-container>\n",
                        changeDetection: core.ChangeDetectionStrategy.OnPush
                    }] }
        ];
        /** @nocollapse */
        DynamicBootstrapRatingComponent.ctorParameters = function () { return [
            { type: core$1.DynamicFormLayoutService },
            { type: core$1.DynamicFormValidationService }
        ]; };
        DynamicBootstrapRatingComponent.propDecorators = {
            group: [{ type: core.Input }],
            layout: [{ type: core.Input }],
            model: [{ type: core.Input }],
            blur: [{ type: core.Output }],
            change: [{ type: core.Output }],
            focus: [{ type: core.Output }]
        };
        return DynamicBootstrapRatingComponent;
    }(core$1.DynamicFormControlComponent));

    var __extends$7 = (this && this.__extends) || (function () {
        var extendStatics = function (d, b) {
            extendStatics = Object.setPrototypeOf ||
                ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
                function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
            return extendStatics(d, b);
        };
        return function (d, b) {
            extendStatics(d, b);
            function __() { this.constructor = d; }
            d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
        };
    })();
    var DynamicBootstrapSelectComponent = /** @class */ (function (_super) {
        __extends$7(DynamicBootstrapSelectComponent, _super);
        function DynamicBootstrapSelectComponent(layoutService, validationService) {
            var _this = _super.call(this, layoutService, validationService) || this;
            _this.layoutService = layoutService;
            _this.validationService = validationService;
            _this.blur = new core.EventEmitter();
            _this.change = new core.EventEmitter();
            _this.focus = new core.EventEmitter();
            return _this;
        }
        DynamicBootstrapSelectComponent.decorators = [
            { type: core.Component, args: [{
                        selector: "dynamic-bootstrap-select",
                        template: "<ng-container [formGroup]=\"group\">\n\n    <select class=\"form-control\"\n            [compareWith]=\"model.compareWithFn\"\n            [formControlName]=\"model.id\"\n            [id]=\"elementId\"\n            [name]=\"model.name\"\n            [ngClass]=\"getClass('element', 'control')\"\n            [required]=\"model.required\"\n            [tabindex]=\"model.tabIndex\"\n            (blur)=\"onBlur($event)\"\n            (change)=\"onChange($event)\"\n            (focus)=\"onFocus($event)\">\n\n        <option *ngFor=\"let option of model.options$ | async\"\n                [disabled]=\"option.disabled\"\n                [ngValue]=\"option.value\">{{ option.label }}\n        </option>\n\n    </select>\n\n</ng-container>\n",
                        changeDetection: core.ChangeDetectionStrategy.OnPush
                    }] }
        ];
        /** @nocollapse */
        DynamicBootstrapSelectComponent.ctorParameters = function () { return [
            { type: core$1.DynamicFormLayoutService },
            { type: core$1.DynamicFormValidationService }
        ]; };
        DynamicBootstrapSelectComponent.propDecorators = {
            group: [{ type: core.Input }],
            layout: [{ type: core.Input }],
            model: [{ type: core.Input }],
            blur: [{ type: core.Output }],
            change: [{ type: core.Output }],
            focus: [{ type: core.Output }]
        };
        return DynamicBootstrapSelectComponent;
    }(core$1.DynamicFormControlComponent));

    var __extends$8 = (this && this.__extends) || (function () {
        var extendStatics = function (d, b) {
            extendStatics = Object.setPrototypeOf ||
                ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
                function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
            return extendStatics(d, b);
        };
        return function (d, b) {
            extendStatics(d, b);
            function __() { this.constructor = d; }
            d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
        };
    })();
    var DynamicBootstrapTextAreaComponent = /** @class */ (function (_super) {
        __extends$8(DynamicBootstrapTextAreaComponent, _super);
        function DynamicBootstrapTextAreaComponent(layoutService, validationService) {
            var _this = _super.call(this, layoutService, validationService) || this;
            _this.layoutService = layoutService;
            _this.validationService = validationService;
            _this.blur = new core.EventEmitter();
            _this.change = new core.EventEmitter();
            _this.focus = new core.EventEmitter();
            return _this;
        }
        DynamicBootstrapTextAreaComponent.decorators = [
            { type: core.Component, args: [{
                        selector: "dynamic-bootstrap-textarea",
                        template: "<ng-container [formGroup]=\"group\">\n\n    <textarea class=\"form-control\"\n              [cols]=\"model.cols\"\n              [formControlName]=\"elementId\"\n              [id]=\"elementId\"\n              [maxlength]=\"model.maxLength\"\n              [minlength]=\"model.minLength\"\n              [name]=\"model.name\"\n              [ngClass]=\"getClass('element', 'control')\"\n              [placeholder]=\"model.placeholder\"\n              [readonly]=\"model.readOnly\"\n              [required]=\"model.required\"\n              [rows]=\"model.rows\"\n              [spellcheck]=\"model.spellCheck\"\n              [tabindex]=\"model.tabIndex\"\n              [wrap]=\"model.wrap\"\n              (blur)=\"onBlur($event)\"\n              (change)=\"onChange($event)\"\n              (focus)=\"onFocus($event)\"></textarea>\n\n</ng-container>\n",
                        changeDetection: core.ChangeDetectionStrategy.OnPush
                    }] }
        ];
        /** @nocollapse */
        DynamicBootstrapTextAreaComponent.ctorParameters = function () { return [
            { type: core$1.DynamicFormLayoutService },
            { type: core$1.DynamicFormValidationService }
        ]; };
        DynamicBootstrapTextAreaComponent.propDecorators = {
            group: [{ type: core.Input }],
            layout: [{ type: core.Input }],
            model: [{ type: core.Input }],
            blur: [{ type: core.Output }],
            change: [{ type: core.Output }],
            focus: [{ type: core.Output }]
        };
        return DynamicBootstrapTextAreaComponent;
    }(core$1.DynamicFormControlComponent));

    var __extends$9 = (this && this.__extends) || (function () {
        var extendStatics = function (d, b) {
            extendStatics = Object.setPrototypeOf ||
                ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
                function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
            return extendStatics(d, b);
        };
        return function (d, b) {
            extendStatics(d, b);
            function __() { this.constructor = d; }
            d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
        };
    })();
    var DynamicBootstrapTimePickerComponent = /** @class */ (function (_super) {
        __extends$9(DynamicBootstrapTimePickerComponent, _super);
        function DynamicBootstrapTimePickerComponent(layoutService, validationService) {
            var _this = _super.call(this, layoutService, validationService) || this;
            _this.layoutService = layoutService;
            _this.validationService = validationService;
            _this.blur = new core.EventEmitter();
            _this.change = new core.EventEmitter();
            _this.focus = new core.EventEmitter();
            return _this;
        }
        DynamicBootstrapTimePickerComponent.decorators = [
            { type: core.Component, args: [{
                        selector: "dynamic-bootstrap-timepicker",
                        template: "<ng-container [formGroup]=\"group\">\n\n    <timepicker [arrowkeys]=\"model.getAdditional('arrowkeys', true)\"\n                [formControlName]=\"model.id\"\n                [hourStep]=\"model.getAdditional('hourStep', 1)\"\n                [id]=\"elementId\"\n                [max]=\"model.max\"\n                [min]=\"model.min\"\n                [mousewheel]=\"model.getAdditional('mousewheel', true)\"\n                [minuteStep]=\"model.getAdditional('minuteStep', 1)\"\n                [ngClass]=\"getClass('element', 'control')\"\n                [secondsStep]=\"model.getAdditional('secondsStep', 1)\"\n                [showMeridian]=\"model.meridian\"\n                [showSeconds]=\"model.showSeconds\"\n                [showSpinners]=\"model.getAdditional('spinners', true)\"></timepicker>\n\n</ng-container>\n",
                        changeDetection: core.ChangeDetectionStrategy.OnPush
                    }] }
        ];
        /** @nocollapse */
        DynamicBootstrapTimePickerComponent.ctorParameters = function () { return [
            { type: core$1.DynamicFormLayoutService },
            { type: core$1.DynamicFormValidationService }
        ]; };
        DynamicBootstrapTimePickerComponent.propDecorators = {
            group: [{ type: core.Input }],
            layout: [{ type: core.Input }],
            model: [{ type: core.Input }],
            blur: [{ type: core.Output }],
            change: [{ type: core.Output }],
            focus: [{ type: core.Output }],
            bsTimePicker: [{ type: core.ViewChild, args: [timepicker.TimepickerComponent,] }]
        };
        return DynamicBootstrapTimePickerComponent;
    }(core$1.DynamicFormControlComponent));

    var __extends$a = (this && this.__extends) || (function () {
        var extendStatics = function (d, b) {
            extendStatics = Object.setPrototypeOf ||
                ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
                function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
            return extendStatics(d, b);
        };
        return function (d, b) {
            extendStatics(d, b);
            function __() { this.constructor = d; }
            d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
        };
    })();
    var DynamicBootstrapFormControlContainerComponent = /** @class */ (function (_super) {
        __extends$a(DynamicBootstrapFormControlContainerComponent, _super);
        function DynamicBootstrapFormControlContainerComponent(componentFactoryResolver, layoutService, validationService, dynamicFormInstancesService) {
            var _this = _super.call(this, componentFactoryResolver, layoutService, validationService, dynamicFormInstancesService) || this;
            _this.componentFactoryResolver = componentFactoryResolver;
            _this.layoutService = layoutService;
            _this.validationService = validationService;
            _this.dynamicFormInstancesService = dynamicFormInstancesService;
            _this.asBootstrapFormGroup = true;
            _this.context = null;
            _this.blur = new core.EventEmitter();
            _this.change = new core.EventEmitter();
            _this.focus = new core.EventEmitter();
            _this.customEvent = new core.EventEmitter();
            return _this;
        }
        Object.defineProperty(DynamicBootstrapFormControlContainerComponent.prototype, "componentType", {
            get: /**
             * @return {?}
             */
            function () {
                return this.layoutService.getCustomComponentType(this.model) || bootstrapUIFormControlMapFn(this.model);
            },
            enumerable: true,
            configurable: true
        });
        DynamicBootstrapFormControlContainerComponent.decorators = [
            { type: core.Component, args: [{
                        selector: "dynamic-bootstrap-form-control",
                        template: "<div [class.form-group]=\"asBootstrapFormGroup || getClass('element', 'container').includes('form-group')\"\n     [class.has-error]=\"showErrorMessages\"\n     [formGroup]=\"group\"\n     [ngClass]=\"[getClass('element', 'container'), getClass('grid', 'container')]\">\n\n    <label *ngIf=\"!isCheckbox && hasLabel\"\n           [for]=\"elementId\"\n           [innerHTML]=\"model.label\"\n           [ngClass]=\"[getClass('element', 'label'), getClass('grid', 'label')]\"></label>\n\n    <ng-container *ngTemplateOutlet=\"startTemplate?.templateRef; context: model\"></ng-container>\n\n    <div [ngClass]=\"getClass('grid', 'control')\">\n\n        <ng-container #componentViewContainer></ng-container>\n\n        <span *ngIf=\"hasHint\" class=\"help-block\" [innerHTML]=\"hint\"\n              [ngClass]=\"getClass('element', 'hint')\"></span>\n\n    </div>\n\n    <ul *ngIf=\"showErrorMessages\" class=\"has-error\"\n        [ngClass]=\"[getClass('element', 'errors'), getClass('grid', 'errors')]\">\n\n        <li *ngFor=\"let message of errorMessages\" class=\"help-block\">{{ message }}</li>\n\n    </ul>\n\n    <ng-container *ngTemplateOutlet=\"endTemplate?.templateRef; context: model\"></ng-container>\n\n    <ng-content></ng-content>\n\n</div>\n"
                    }] }
        ];
        /** @nocollapse */
        DynamicBootstrapFormControlContainerComponent.ctorParameters = function () { return [
            { type: core.ComponentFactoryResolver },
            { type: core$1.DynamicFormLayoutService },
            { type: core$1.DynamicFormValidationService },
            { type: core$1.DynamicFormInstancesService }
        ]; };
        DynamicBootstrapFormControlContainerComponent.propDecorators = {
            contentTemplateList: [{ type: core.ContentChildren, args: [core$1.DynamicTemplateDirective,] }],
            inputTemplateList: [{ type: core.Input, args: ["templates",] }],
            asBootstrapFormGroup: [{ type: core.Input }],
            context: [{ type: core.Input }],
            group: [{ type: core.Input }],
            layout: [{ type: core.Input }],
            model: [{ type: core.Input }],
            blur: [{ type: core.Output }],
            change: [{ type: core.Output }],
            focus: [{ type: core.Output }],
            customEvent: [{ type: core.Output, args: ["bsEvent",] }],
            componentViewContainerRef: [{ type: core.ViewChild, args: ["componentViewContainer", { read: core.ViewContainerRef },] }]
        };
        return DynamicBootstrapFormControlContainerComponent;
    }(core$1.DynamicFormControlContainerComponent));
    /**
     * @param {?} model
     * @return {?}
     */
    function bootstrapUIFormControlMapFn(model) {
        switch (model.type) {
            case core$1.DYNAMIC_FORM_CONTROL_TYPE_ARRAY:
                return DynamicBootstrapFormArrayComponent;
            case core$1.DYNAMIC_FORM_CONTROL_TYPE_CHECKBOX:
                return DynamicBootstrapCheckboxComponent;
            case core$1.DYNAMIC_FORM_CONTROL_TYPE_CHECKBOX_GROUP:
                return DynamicBootstrapFormGroupComponent;
            case core$1.DYNAMIC_FORM_CONTROL_TYPE_DATEPICKER:
                return DynamicBootstrapDatePickerComponent;
            case core$1.DYNAMIC_FORM_CONTROL_TYPE_GROUP:
                return DynamicBootstrapFormGroupComponent;
            case core$1.DYNAMIC_FORM_CONTROL_TYPE_INPUT:
                return DynamicBootstrapInputComponent;
            case core$1.DYNAMIC_FORM_CONTROL_TYPE_RADIO_GROUP:
                return DynamicBootstrapRadioGroupComponent;
            case core$1.DYNAMIC_FORM_CONTROL_TYPE_RATING:
                return DynamicBootstrapRatingComponent;
            case core$1.DYNAMIC_FORM_CONTROL_TYPE_SELECT:
                return DynamicBootstrapSelectComponent;
            case core$1.DYNAMIC_FORM_CONTROL_TYPE_TEXTAREA:
                return DynamicBootstrapTextAreaComponent;
            case core$1.DYNAMIC_FORM_CONTROL_TYPE_TIMEPICKER:
                return DynamicBootstrapTimePickerComponent;
            default:
                return null;
        }
    }

    var __extends$b = (this && this.__extends) || (function () {
        var extendStatics = function (d, b) {
            extendStatics = Object.setPrototypeOf ||
                ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
                function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
            return extendStatics(d, b);
        };
        return function (d, b) {
            extendStatics(d, b);
            function __() { this.constructor = d; }
            d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
        };
    })();
    var DynamicBootstrapFormComponent = /** @class */ (function (_super) {
        __extends$b(DynamicBootstrapFormComponent, _super);
        function DynamicBootstrapFormComponent(formService, layoutService) {
            var _this = _super.call(this, formService, layoutService) || this;
            _this.formService = formService;
            _this.layoutService = layoutService;
            _this.blur = new core.EventEmitter();
            _this.change = new core.EventEmitter();
            _this.focus = new core.EventEmitter();
            _this.customEvent = new core.EventEmitter();
            return _this;
        }
        DynamicBootstrapFormComponent.decorators = [
            { type: core.Component, args: [{
                        selector: "dynamic-bootstrap-form",
                        template: "<dynamic-bootstrap-form-control *ngFor=\"let model of formModel; trackBy: trackByFn\"\n                                [group]=\"formGroup\"\n                                [hidden]=\"model.hidden\"\n                                [layout]=\"formLayout\"\n                                [model]=\"model\"\n                                [ngClass]=\"[getClass(model,'element','host'), getClass(model,'grid','host')]\"\n                                [templates]=\"templates\"\n                                (blur)=\"onEvent($event, 'blur')\"\n                                (change)=\"onEvent($event, 'change')\"\n                                (focus)=\"onEvent($event, 'focus')\"\n                                (bsEvent)=\"onEvent($event, 'custom')\"></dynamic-bootstrap-form-control>"
                    }] }
        ];
        /** @nocollapse */
        DynamicBootstrapFormComponent.ctorParameters = function () { return [
            { type: core$1.DynamicFormService },
            { type: core$1.DynamicFormLayoutService }
        ]; };
        DynamicBootstrapFormComponent.propDecorators = {
            formGroup: [{ type: core.Input, args: ["group",] }],
            formModel: [{ type: core.Input, args: ["model",] }],
            formLayout: [{ type: core.Input, args: ["layout",] }],
            blur: [{ type: core.Output }],
            change: [{ type: core.Output }],
            focus: [{ type: core.Output }],
            customEvent: [{ type: core.Output, args: ["bsEvent",] }],
            templates: [{ type: core.ContentChildren, args: [core$1.DynamicTemplateDirective,] }],
            components: [{ type: core.ViewChildren, args: [DynamicBootstrapFormControlContainerComponent,] }]
        };
        return DynamicBootstrapFormComponent;
    }(core$1.DynamicFormComponent));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var /** @type {?} */ RATING_CONTROL_VALUE_ACCESSOR = {
        provide: forms.NG_VALUE_ACCESSOR,
        /* tslint:disable-next-line: no-use-before-declare */
        useExisting: core.forwardRef(function () { return RatingComponent; }),
        multi: true
    };
    var RatingComponent = /** @class */ (function () {
        function RatingComponent(changeDetection) {
            this.changeDetection = changeDetection;
            /**
             * number of icons
             */
            this.max = 5;
            /**
             * fired when icon selected, $event:number equals to selected rating
             */
            this.onHover = new core.EventEmitter();
            /**
             * fired when icon selected, $event:number equals to previous rating value
             */
            this.onLeave = new core.EventEmitter();
            // tslint:disable-next-line:no-any
            this.onChange = Function.prototype;
            // tslint:disable-next-line:no-any
            this.onTouched = Function.prototype;
        }
        /**
         * @param {?} event
         * @return {?}
         */
        RatingComponent.prototype.onKeydown = /**
         * @param {?} event
         * @return {?}
         */
        function (event) {
            /* tslint:disable-next-line: deprecation */
            if ([37, 38, 39, 40].indexOf(event.which) === -1) {
                return;
            }
            event.preventDefault();
            event.stopPropagation();
            /* tslint:disable-next-line: deprecation */
            var /** @type {?} */ sign = event.which === 38 || event.which === 39 ? 1 : -1;
            this.rate(this.value + sign);
        };
        /**
         * @return {?}
         */
        RatingComponent.prototype.ngOnInit = /**
         * @return {?}
         */
        function () {
            this.max = typeof this.max !== 'undefined' ? this.max : 5;
            this.titles =
                typeof this.titles !== 'undefined' && this.titles.length > 0
                    ? this.titles
                    : [];
            this.range = this.buildTemplateObjects(this.max);
        };
        // model -> view
        /**
         * @param {?} value
         * @return {?}
         */
        RatingComponent.prototype.writeValue = /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            if (value % 1 !== value) {
                this.value = Math.round(value);
                this.preValue = value;
                this.changeDetection.markForCheck();
                return;
            }
            this.preValue = value;
            this.value = value;
            this.changeDetection.markForCheck();
        };
        /**
         * @param {?} value
         * @return {?}
         */
        RatingComponent.prototype.enter = /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            if (!this.readonly) {
                this.value = value;
                this.changeDetection.markForCheck();
                this.onHover.emit(value);
            }
        };
        /**
         * @return {?}
         */
        RatingComponent.prototype.reset = /**
         * @return {?}
         */
        function () {
            this.value = this.preValue;
            this.changeDetection.markForCheck();
            this.onLeave.emit(this.value);
        };
        /**
         * @param {?} fn
         * @return {?}
         */
        RatingComponent.prototype.registerOnChange = /**
         * @param {?} fn
         * @return {?}
         */
        function (fn) {
            this.onChange = fn;
        };
        /**
         * @param {?} fn
         * @return {?}
         */
        RatingComponent.prototype.registerOnTouched = /**
         * @param {?} fn
         * @return {?}
         */
        function (fn) {
            this.onTouched = fn;
        };
        /**
         * @param {?} value
         * @return {?}
         */
        RatingComponent.prototype.rate = /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            if (!this.readonly && value >= 0 && value <= this.range.length) {
                this.writeValue(value);
                this.onChange(value);
            }
        };
        /**
         * @param {?} max
         * @return {?}
         */
        RatingComponent.prototype.buildTemplateObjects = /**
         * @param {?} max
         * @return {?}
         */
        function (max) {
            var /** @type {?} */ result = [];
            for (var /** @type {?} */ i = 0; i < max; i++) {
                result.push({
                    index: i,
                    title: this.titles[i] || i + 1
                });
            }
            return result;
        };
        RatingComponent.decorators = [
            { type: core.Component, args: [{
                        selector: 'rating',
                        template: "<span (mouseleave)=\"reset()\" (keydown)=\"onKeydown($event)\" tabindex=\"0\"\n      role=\"slider\" aria-valuemin=\"0\" [attr.aria-valuemax]=\"range.length\"\n      [attr.aria-valuenow]=\"value\">\n  <ng-template #star let-value=\"value\" let-index=\"index\">{{ index < value ? '&#9733;' : '&#9734;' }}</ng-template>\n  <ng-template ngFor let-r [ngForOf]=\"range\" let-index=\"index\">\n    <span class=\"sr-only\">({{ index < value ? '*' : ' ' }})</span>\n    <span class=\"bs-rating-star\"\n          (mouseenter)=\"enter(index + 1)\"\n          (click)=\"rate(index + 1)\"\n          [title]=\"r.title\"\n          [style.cursor]=\"readonly ? 'default' : 'pointer'\"\n          [class.active]=\"index < value\">\n      <ng-template [ngTemplateOutlet]=\"customTemplate || star\"\n                   [ngTemplateOutletContext]=\"{index: index, value: value}\">\n      </ng-template>\n    </span>\n  </ng-template>\n</span>\n",
                        providers: [RATING_CONTROL_VALUE_ACCESSOR],
                        changeDetection: core.ChangeDetectionStrategy.OnPush
                    }] }
        ];
        /** @nocollapse */
        RatingComponent.ctorParameters = function () { return [
            { type: core.ChangeDetectorRef, },
        ]; };
        RatingComponent.propDecorators = {
            "max": [{ type: core.Input },],
            "readonly": [{ type: core.Input },],
            "titles": [{ type: core.Input },],
            "customTemplate": [{ type: core.Input },],
            "onHover": [{ type: core.Output },],
            "onLeave": [{ type: core.Output },],
            "onKeydown": [{ type: core.HostListener, args: ['keydown', ['$event'],] },],
        };
        return RatingComponent;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var RatingModule = /** @class */ (function () {
        function RatingModule() {
        }
        /**
         * @return {?}
         */
        RatingModule.forRoot = /**
         * @return {?}
         */
        function () {
            return {
                ngModule: RatingModule,
                providers: []
            };
        };
        RatingModule.decorators = [
            { type: core.NgModule, args: [{
                        imports: [common.CommonModule],
                        declarations: [RatingComponent],
                        exports: [RatingComponent]
                    },] }
        ];
        return RatingModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    var DynamicFormsBootstrapUIModule = /** @class */ (function () {
        function DynamicFormsBootstrapUIModule() {
        }
        DynamicFormsBootstrapUIModule.decorators = [
            { type: core.NgModule, args: [{
                        imports: [
                            common.CommonModule,
                            forms.ReactiveFormsModule,
                            datepicker.BsDatepickerModule,
                            RatingModule,
                            timepicker.TimepickerModule,
                            angular2TextMask.TextMaskModule,
                            core$1.DynamicFormsCoreModule
                        ],
                        declarations: [
                            DynamicBootstrapCheckboxComponent,
                            DynamicBootstrapDatePickerComponent,
                            DynamicBootstrapFormArrayComponent,
                            DynamicBootstrapFormComponent,
                            DynamicBootstrapFormControlContainerComponent,
                            DynamicBootstrapFormGroupComponent,
                            DynamicBootstrapInputComponent,
                            DynamicBootstrapRadioGroupComponent,
                            DynamicBootstrapRatingComponent,
                            DynamicBootstrapSelectComponent,
                            DynamicBootstrapTextAreaComponent,
                            DynamicBootstrapTimePickerComponent
                        ],
                        entryComponents: [
                            DynamicBootstrapCheckboxComponent,
                            DynamicBootstrapDatePickerComponent,
                            DynamicBootstrapFormArrayComponent,
                            DynamicBootstrapFormGroupComponent,
                            DynamicBootstrapInputComponent,
                            DynamicBootstrapRadioGroupComponent,
                            DynamicBootstrapRatingComponent,
                            DynamicBootstrapSelectComponent,
                            DynamicBootstrapTextAreaComponent,
                            DynamicBootstrapTimePickerComponent
                        ],
                        exports: [
                            core$1.DynamicFormsCoreModule,
                            DynamicBootstrapCheckboxComponent,
                            DynamicBootstrapDatePickerComponent,
                            DynamicBootstrapFormArrayComponent,
                            DynamicBootstrapFormComponent,
                            DynamicBootstrapFormControlContainerComponent,
                            DynamicBootstrapFormGroupComponent,
                            DynamicBootstrapInputComponent,
                            DynamicBootstrapRadioGroupComponent,
                            DynamicBootstrapRatingComponent,
                            DynamicBootstrapSelectComponent,
                            DynamicBootstrapTextAreaComponent,
                            DynamicBootstrapTimePickerComponent
                        ]
                    },] }
        ];
        return DynamicFormsBootstrapUIModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    exports.DynamicBootstrapCheckboxComponent = DynamicBootstrapCheckboxComponent;
    exports.DynamicBootstrapDatePickerComponent = DynamicBootstrapDatePickerComponent;
    exports.DynamicBootstrapFormArrayComponent = DynamicBootstrapFormArrayComponent;
    exports.DynamicBootstrapFormGroupComponent = DynamicBootstrapFormGroupComponent;
    exports.DynamicBootstrapInputComponent = DynamicBootstrapInputComponent;
    exports.DynamicBootstrapRadioGroupComponent = DynamicBootstrapRadioGroupComponent;
    exports.DynamicBootstrapRatingComponent = DynamicBootstrapRatingComponent;
    exports.DynamicBootstrapSelectComponent = DynamicBootstrapSelectComponent;
    exports.DynamicBootstrapTextAreaComponent = DynamicBootstrapTextAreaComponent;
    exports.DynamicBootstrapTimePickerComponent = DynamicBootstrapTimePickerComponent;
    exports.DynamicBootstrapFormComponent = DynamicBootstrapFormComponent;
    exports.bootstrapUIFormControlMapFn = bootstrapUIFormControlMapFn;
    exports.DynamicBootstrapFormControlContainerComponent = DynamicBootstrapFormControlContainerComponent;
    exports.DynamicFormsBootstrapUIModule = DynamicFormsBootstrapUIModule;

    Object.defineProperty(exports, '__esModule', { value: true });

}));
