// @ng-dynamic-forms/ui-bootstrap/esm5/src/checkbox/dynamic-bootstrap-checkbox.component.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output } from "@angular/core";
import { FormGroup } from "@angular/forms";
import { DynamicCheckboxModel, DynamicFormControlComponent, DynamicFormLayoutService, DynamicFormValidationService, } from "@ng-dynamic-forms/core";
var DynamicBootstrapCheckboxComponent = /** @class */ (function (_super) {
    __extends(DynamicBootstrapCheckboxComponent, _super);
    function DynamicBootstrapCheckboxComponent(layoutService, validationService) {
        var _this = _super.call(this, layoutService, validationService) || this;
        _this.layoutService = layoutService;
        _this.validationService = validationService;
        _this.blur = new EventEmitter();
        _this.change = new EventEmitter();
        _this.focus = new EventEmitter();
        return _this;
    }
    DynamicBootstrapCheckboxComponent.decorators = [
        { type: Component, args: [{
                    selector: "dynamic-bootstrap-checkbox",
                    template: "<div [formGroup]=\"group\" [class.disabled]=\"model.disabled\" [ngClass]=\"getClass('element', 'control') || 'checkbox'\">\n\n    <label>\n\n        <input type=\"checkbox\"\n               [checked]=\"model.checked\"\n               [formControlName]=\"model.id\"\n               [id]=\"elementId\"\n               [indeterminate]=\"model.indeterminate\"\n               [name]=\"model.name\"\n               [ngClass]=\"getClass('element', 'control')\"\n               [required]=\"model.required\"\n               [tabindex]=\"model.tabIndex\"\n               [value]=\"model.value\"\n               (blur)=\"onBlur($event)\"\n               (change)=\"onChange($event)\"\n               (focus)=\"onFocus($event)\"/><span [innerHTML]=\"model.label\"\n                                                [ngClass]=\"[getClass('element', 'label'), getClass('grid', 'label')]\"></span></label>\n</div>\n",
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    /** @nocollapse */
    DynamicBootstrapCheckboxComponent.ctorParameters = function () { return [
        { type: DynamicFormLayoutService },
        { type: DynamicFormValidationService }
    ]; };
    DynamicBootstrapCheckboxComponent.propDecorators = {
        group: [{ type: Input }],
        layout: [{ type: Input }],
        model: [{ type: Input }],
        blur: [{ type: Output }],
        change: [{ type: Output }],
        focus: [{ type: Output }]
    };
    return DynamicBootstrapCheckboxComponent;
}(DynamicFormControlComponent));
export { DynamicBootstrapCheckboxComponent };
if (false) {
    /** @type {?} */
    DynamicBootstrapCheckboxComponent.prototype.group;
    /** @type {?} */
    DynamicBootstrapCheckboxComponent.prototype.layout;
    /** @type {?} */
    DynamicBootstrapCheckboxComponent.prototype.model;
    /** @type {?} */
    DynamicBootstrapCheckboxComponent.prototype.blur;
    /** @type {?} */
    DynamicBootstrapCheckboxComponent.prototype.change;
    /** @type {?} */
    DynamicBootstrapCheckboxComponent.prototype.focus;
    /**
     * @type {?}
     * @protected
     */
    DynamicBootstrapCheckboxComponent.prototype.layoutService;
    /**
     * @type {?}
     * @protected
     */
    DynamicBootstrapCheckboxComponent.prototype.validationService;
}
