// @ng-dynamic-forms/ui-bootstrap/esm2015/src/timepicker/dynamic-bootstrap-timepicker.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output, ViewChild } from "@angular/core";
import { FormGroup } from "@angular/forms";
import { TimepickerComponent } from "ngx-bootstrap/timepicker";
import { DynamicFormControlComponent, DynamicFormLayoutService, DynamicFormValidationService, DynamicTimePickerModel } from "@ng-dynamic-forms/core";
export class DynamicBootstrapTimePickerComponent extends DynamicFormControlComponent {
    /**
     * @param {?} layoutService
     * @param {?} validationService
     */
    constructor(layoutService, validationService) {
        super(layoutService, validationService);
        this.layoutService = layoutService;
        this.validationService = validationService;
        this.blur = new EventEmitter();
        this.change = new EventEmitter();
        this.focus = new EventEmitter();
    }
}
DynamicBootstrapTimePickerComponent.decorators = [
    { type: Component, args: [{
                selector: "dynamic-bootstrap-timepicker",
                template: "<ng-container [formGroup]=\"group\">\n\n    <timepicker [arrowkeys]=\"model.getAdditional('arrowkeys', true)\"\n                [formControlName]=\"model.id\"\n                [hourStep]=\"model.getAdditional('hourStep', 1)\"\n                [id]=\"elementId\"\n                [max]=\"model.max\"\n                [min]=\"model.min\"\n                [mousewheel]=\"model.getAdditional('mousewheel', true)\"\n                [minuteStep]=\"model.getAdditional('minuteStep', 1)\"\n                [ngClass]=\"getClass('element', 'control')\"\n                [secondsStep]=\"model.getAdditional('secondsStep', 1)\"\n                [showMeridian]=\"model.meridian\"\n                [showSeconds]=\"model.showSeconds\"\n                [showSpinners]=\"model.getAdditional('spinners', true)\"></timepicker>\n\n</ng-container>\n",
                changeDetection: ChangeDetectionStrategy.OnPush
            }] }
];
/** @nocollapse */
DynamicBootstrapTimePickerComponent.ctorParameters = () => [
    { type: DynamicFormLayoutService },
    { type: DynamicFormValidationService }
];
DynamicBootstrapTimePickerComponent.propDecorators = {
    group: [{ type: Input }],
    layout: [{ type: Input }],
    model: [{ type: Input }],
    blur: [{ type: Output }],
    change: [{ type: Output }],
    focus: [{ type: Output }],
    bsTimePicker: [{ type: ViewChild, args: [TimepickerComponent,] }]
};
if (false) {
    /** @type {?} */
    DynamicBootstrapTimePickerComponent.prototype.group;
    /** @type {?} */
    DynamicBootstrapTimePickerComponent.prototype.layout;
    /** @type {?} */
    DynamicBootstrapTimePickerComponent.prototype.model;
    /** @type {?} */
    DynamicBootstrapTimePickerComponent.prototype.blur;
    /** @type {?} */
    DynamicBootstrapTimePickerComponent.prototype.change;
    /** @type {?} */
    DynamicBootstrapTimePickerComponent.prototype.focus;
    /** @type {?} */
    DynamicBootstrapTimePickerComponent.prototype.bsTimePicker;
    /**
     * @type {?}
     * @protected
     */
    DynamicBootstrapTimePickerComponent.prototype.layoutService;
    /**
     * @type {?}
     * @protected
     */
    DynamicBootstrapTimePickerComponent.prototype.validationService;
}
