// @ng-dynamic-forms/ui-bootstrap/esm5/src/timepicker/dynamic-bootstrap-timepicker.component.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output, ViewChild } from "@angular/core";
import { FormGroup } from "@angular/forms";
import { TimepickerComponent } from "ngx-bootstrap/timepicker";
import { DynamicFormControlComponent, DynamicFormLayoutService, DynamicFormValidationService, DynamicTimePickerModel } from "@ng-dynamic-forms/core";
var DynamicBootstrapTimePickerComponent = /** @class */ (function (_super) {
    __extends(DynamicBootstrapTimePickerComponent, _super);
    function DynamicBootstrapTimePickerComponent(layoutService, validationService) {
        var _this = _super.call(this, layoutService, validationService) || this;
        _this.layoutService = layoutService;
        _this.validationService = validationService;
        _this.blur = new EventEmitter();
        _this.change = new EventEmitter();
        _this.focus = new EventEmitter();
        return _this;
    }
    DynamicBootstrapTimePickerComponent.decorators = [
        { type: Component, args: [{
                    selector: "dynamic-bootstrap-timepicker",
                    template: "<ng-container [formGroup]=\"group\">\n\n    <timepicker [arrowkeys]=\"model.getAdditional('arrowkeys', true)\"\n                [formControlName]=\"model.id\"\n                [hourStep]=\"model.getAdditional('hourStep', 1)\"\n                [id]=\"elementId\"\n                [max]=\"model.max\"\n                [min]=\"model.min\"\n                [mousewheel]=\"model.getAdditional('mousewheel', true)\"\n                [minuteStep]=\"model.getAdditional('minuteStep', 1)\"\n                [ngClass]=\"getClass('element', 'control')\"\n                [secondsStep]=\"model.getAdditional('secondsStep', 1)\"\n                [showMeridian]=\"model.meridian\"\n                [showSeconds]=\"model.showSeconds\"\n                [showSpinners]=\"model.getAdditional('spinners', true)\"></timepicker>\n\n</ng-container>\n",
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    /** @nocollapse */
    DynamicBootstrapTimePickerComponent.ctorParameters = function () { return [
        { type: DynamicFormLayoutService },
        { type: DynamicFormValidationService }
    ]; };
    DynamicBootstrapTimePickerComponent.propDecorators = {
        group: [{ type: Input }],
        layout: [{ type: Input }],
        model: [{ type: Input }],
        blur: [{ type: Output }],
        change: [{ type: Output }],
        focus: [{ type: Output }],
        bsTimePicker: [{ type: ViewChild, args: [TimepickerComponent,] }]
    };
    return DynamicBootstrapTimePickerComponent;
}(DynamicFormControlComponent));
export { DynamicBootstrapTimePickerComponent };
if (false) {
    /** @type {?} */
    DynamicBootstrapTimePickerComponent.prototype.group;
    /** @type {?} */
    DynamicBootstrapTimePickerComponent.prototype.layout;
    /** @type {?} */
    DynamicBootstrapTimePickerComponent.prototype.model;
    /** @type {?} */
    DynamicBootstrapTimePickerComponent.prototype.blur;
    /** @type {?} */
    DynamicBootstrapTimePickerComponent.prototype.change;
    /** @type {?} */
    DynamicBootstrapTimePickerComponent.prototype.focus;
    /** @type {?} */
    DynamicBootstrapTimePickerComponent.prototype.bsTimePicker;
    /**
     * @type {?}
     * @protected
     */
    DynamicBootstrapTimePickerComponent.prototype.layoutService;
    /**
     * @type {?}
     * @protected
     */
    DynamicBootstrapTimePickerComponent.prototype.validationService;
}
