// @ng-dynamic-forms/ui-bootstrap/esm5/src/form-group/dynamic-bootstrap-form-group.component.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from "@angular/core";
import { FormGroup } from "@angular/forms";
import { DynamicFormControlComponent, DynamicFormGroupModel, DynamicFormLayoutService, DynamicFormValidationService } from "@ng-dynamic-forms/core";
var DynamicBootstrapFormGroupComponent = /** @class */ (function (_super) {
    __extends(DynamicBootstrapFormGroupComponent, _super);
    function DynamicBootstrapFormGroupComponent(layoutService, validationService) {
        var _this = _super.call(this, layoutService, validationService) || this;
        _this.layoutService = layoutService;
        _this.validationService = validationService;
        _this.blur = new EventEmitter();
        _this.change = new EventEmitter();
        _this.customEvent = new EventEmitter();
        _this.focus = new EventEmitter();
        return _this;
    }
    DynamicBootstrapFormGroupComponent.decorators = [
        { type: Component, args: [{
                    selector: "dynamic-bootstrap-form-group",
                    template: "<ng-container [formGroup]=\"group\">\n\n    <fieldset [formGroupName]=\"model.id\"\n              [id]=\"elementId\"\n              [name]=\"model.name\"\n              [ngClass]=\"getClass('element', 'control')\">\n\n        <legend *ngIf=\"model.legend\" [innerHTML]=\"model.legend\"></legend>\n\n        <dynamic-bootstrap-form-control *ngFor=\"let _model of model.group\"\n                                        [asBootstrapFormGroup]=\"false\"\n                                        [group]=\"control\"\n                                        [hidden]=\"_model.hidden\"\n                                        [layout]=\"layout\"\n                                        [model]=\"_model\"\n                                        [templates]=\"templates\"\n                                        [ngClass]=\"[getClass('element', 'host', _model), getClass('grid', 'host', _model),\n                                                    getClass('element','children'), getClass('grid','children')]\"\n                                        (blur)=\"onBlur($event)\"\n                                        (change)=\"onChange($event)\"\n                                        (focus)=\"onFocus($event)\"\n                                        (bsEvent)=\"onCustomEvent($event, null, true)\"></dynamic-bootstrap-form-control>\n    </fieldset>\n\n</ng-container>\n"
                }] }
    ];
    /** @nocollapse */
    DynamicBootstrapFormGroupComponent.ctorParameters = function () { return [
        { type: DynamicFormLayoutService },
        { type: DynamicFormValidationService }
    ]; };
    DynamicBootstrapFormGroupComponent.propDecorators = {
        group: [{ type: Input }],
        layout: [{ type: Input }],
        model: [{ type: Input }],
        templates: [{ type: Input }],
        blur: [{ type: Output }],
        change: [{ type: Output }],
        customEvent: [{ type: Output }],
        focus: [{ type: Output }]
    };
    return DynamicBootstrapFormGroupComponent;
}(DynamicFormControlComponent));
export { DynamicBootstrapFormGroupComponent };
if (false) {
    /** @type {?} */
    DynamicBootstrapFormGroupComponent.prototype.group;
    /** @type {?} */
    DynamicBootstrapFormGroupComponent.prototype.layout;
    /** @type {?} */
    DynamicBootstrapFormGroupComponent.prototype.model;
    /** @type {?} */
    DynamicBootstrapFormGroupComponent.prototype.templates;
    /** @type {?} */
    DynamicBootstrapFormGroupComponent.prototype.blur;
    /** @type {?} */
    DynamicBootstrapFormGroupComponent.prototype.change;
    /** @type {?} */
    DynamicBootstrapFormGroupComponent.prototype.customEvent;
    /** @type {?} */
    DynamicBootstrapFormGroupComponent.prototype.focus;
    /**
     * @type {?}
     * @protected
     */
    DynamicBootstrapFormGroupComponent.prototype.layoutService;
    /**
     * @type {?}
     * @protected
     */
    DynamicBootstrapFormGroupComponent.prototype.validationService;
}
