// @ng-dynamic-forms/ui-bootstrap/esm5/src/radio-group/dynamic-bootstrap-radio-group.component.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output } from "@angular/core";
import { FormGroup } from "@angular/forms";
import { DynamicFormControlComponent, DynamicFormLayoutService, DynamicFormValidationService, DynamicRadioGroupModel } from "@ng-dynamic-forms/core";
var DynamicBootstrapRadioGroupComponent = /** @class */ (function (_super) {
    __extends(DynamicBootstrapRadioGroupComponent, _super);
    function DynamicBootstrapRadioGroupComponent(layoutService, validationService) {
        var _this = _super.call(this, layoutService, validationService) || this;
        _this.layoutService = layoutService;
        _this.validationService = validationService;
        _this.blur = new EventEmitter();
        _this.change = new EventEmitter();
        _this.focus = new EventEmitter();
        return _this;
    }
    DynamicBootstrapRadioGroupComponent.decorators = [
        { type: Component, args: [{
                    selector: "dynamic-bootstrap-radio-group",
                    template: "<fieldset role=\"radiogroup\"\n          [formGroup]=\"group\"\n          [id]=\"elementId\"\n          [name]=\"model.name\"\n          [ngClass]=\"getClass('element', 'control')\"\n          [tabindex]=\"model.tabIndex\"\n          (change)=\"onChange($event)\">\n\n    <legend *ngIf=\"model.legend\" [innerHTML]=\"model.legend\"></legend>\n\n    <div *ngFor=\"let option of model.options$ | async\" [ngClass]=\"getClass('element', 'option') || 'radio'\">\n\n        <label><input type=\"radio\"\n                      [formControlName]=\"model.id\"\n                      [name]=\"model.name\"\n                      [value]=\"option.value\"\n                      (blur)=\"onBlur($event)\"\n                      (focus)=\"onFocus($event)\"/><span [innerHTML]=\"option.label\"></span></label>\n    </div>\n\n</fieldset>\n",
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    /** @nocollapse */
    DynamicBootstrapRadioGroupComponent.ctorParameters = function () { return [
        { type: DynamicFormLayoutService },
        { type: DynamicFormValidationService }
    ]; };
    DynamicBootstrapRadioGroupComponent.propDecorators = {
        group: [{ type: Input }],
        layout: [{ type: Input }],
        model: [{ type: Input }],
        blur: [{ type: Output }],
        change: [{ type: Output }],
        focus: [{ type: Output }]
    };
    return DynamicBootstrapRadioGroupComponent;
}(DynamicFormControlComponent));
export { DynamicBootstrapRadioGroupComponent };
if (false) {
    /** @type {?} */
    DynamicBootstrapRadioGroupComponent.prototype.group;
    /** @type {?} */
    DynamicBootstrapRadioGroupComponent.prototype.layout;
    /** @type {?} */
    DynamicBootstrapRadioGroupComponent.prototype.model;
    /** @type {?} */
    DynamicBootstrapRadioGroupComponent.prototype.blur;
    /** @type {?} */
    DynamicBootstrapRadioGroupComponent.prototype.change;
    /** @type {?} */
    DynamicBootstrapRadioGroupComponent.prototype.focus;
    /**
     * @type {?}
     * @protected
     */
    DynamicBootstrapRadioGroupComponent.prototype.layoutService;
    /**
     * @type {?}
     * @protected
     */
    DynamicBootstrapRadioGroupComponent.prototype.validationService;
}
