// @ng-dynamic-forms/ui-bootstrap/esm5/src/form-array/dynamic-bootstrap-form-array.component.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from "@angular/core";
import { FormGroup } from "@angular/forms";
import { DynamicFormArrayComponent, DynamicFormArrayModel, DynamicFormLayoutService, DynamicFormValidationService } from "@ng-dynamic-forms/core";
var DynamicBootstrapFormArrayComponent = /** @class */ (function (_super) {
    __extends(DynamicBootstrapFormArrayComponent, _super);
    function DynamicBootstrapFormArrayComponent(layoutService, validationService) {
        var _this = _super.call(this, layoutService, validationService) || this;
        _this.layoutService = layoutService;
        _this.validationService = validationService;
        _this.blur = new EventEmitter();
        _this.change = new EventEmitter();
        _this.customEvent = new EventEmitter();
        _this.focus = new EventEmitter();
        return _this;
    }
    DynamicBootstrapFormArrayComponent.decorators = [
        { type: Component, args: [{
                    selector: "dynamic-bootstrap-form-array",
                    template: "<ng-container [formGroup]=\"group\">\n\n    <div [formArrayName]=\"model.id\" [id]=\"elementId\" [ngClass]=\"getClass('element', 'control')\">\n\n        <div *ngFor=\"let groupModel of model.groups; let idx = index\" role=\"group\"\n             [formGroupName]=\"idx\" [ngClass]=\"[getClass('element', 'group'), getClass('grid', 'group')]\">\n\n            <ng-container *ngTemplateOutlet=\"startTemplate?.templateRef; context: groupModel\"></ng-container>\n\n            <dynamic-bootstrap-form-control *ngFor=\"let _model of groupModel.group\"\n                                            [context]=\"groupModel\"\n                                            [group]=\"array.at(idx)\"\n                                            [hidden]=\"_model.hidden\"\n                                            [layout]=\"layout\"\n                                            [model]=\"_model\"\n                                            [templates]=\"templates\"\n                                            [ngClass]=\"[getClass('element', 'host', _model), getClass('grid', 'host', _model),\n                                                        getClass('element','children'), getClass('grid','children')]\"\n                                            (blur)=\"onBlur($event)\"\n                                            (change)=\"onChange($event)\"\n                                            (focus)=\"onFocus($event)\"\n                                            (bsEvent)=\"onCustomEvent($event, null, true)\"></dynamic-bootstrap-form-control>\n\n            <ng-container *ngTemplateOutlet=\"endTemplate?.templateRef; context: groupModel\"></ng-container>\n\n        </div>\n\n    </div>\n\n</ng-container>\n"
                }] }
    ];
    /** @nocollapse */
    DynamicBootstrapFormArrayComponent.ctorParameters = function () { return [
        { type: DynamicFormLayoutService },
        { type: DynamicFormValidationService }
    ]; };
    DynamicBootstrapFormArrayComponent.propDecorators = {
        group: [{ type: Input }],
        layout: [{ type: Input }],
        model: [{ type: Input }],
        templates: [{ type: Input }],
        blur: [{ type: Output }],
        change: [{ type: Output }],
        customEvent: [{ type: Output }],
        focus: [{ type: Output }]
    };
    return DynamicBootstrapFormArrayComponent;
}(DynamicFormArrayComponent));
export { DynamicBootstrapFormArrayComponent };
if (false) {
    /** @type {?} */
    DynamicBootstrapFormArrayComponent.prototype.group;
    /** @type {?} */
    DynamicBootstrapFormArrayComponent.prototype.layout;
    /** @type {?} */
    DynamicBootstrapFormArrayComponent.prototype.model;
    /** @type {?} */
    DynamicBootstrapFormArrayComponent.prototype.templates;
    /** @type {?} */
    DynamicBootstrapFormArrayComponent.prototype.blur;
    /** @type {?} */
    DynamicBootstrapFormArrayComponent.prototype.change;
    /** @type {?} */
    DynamicBootstrapFormArrayComponent.prototype.customEvent;
    /** @type {?} */
    DynamicBootstrapFormArrayComponent.prototype.focus;
    /**
     * @type {?}
     * @protected
     */
    DynamicBootstrapFormArrayComponent.prototype.layoutService;
    /**
     * @type {?}
     * @protected
     */
    DynamicBootstrapFormArrayComponent.prototype.validationService;
}
