// @ng-dynamic-forms/ui-bootstrap/esm2015/src/datepicker/dynamic-bootstrap-datepicker.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output, ViewChild } from "@angular/core";
import { FormGroup } from "@angular/forms";
import { BsDatepickerDirective } from "ngx-bootstrap/datepicker";
import { DynamicDatePickerModel, DynamicFormControlComponent, DynamicFormLayoutService, DynamicFormValidationService } from "@ng-dynamic-forms/core";
export class DynamicBootstrapDatePickerComponent extends DynamicFormControlComponent {
    /**
     * @param {?} layoutService
     * @param {?} validationService
     */
    constructor(layoutService, validationService) {
        super(layoutService, validationService);
        this.layoutService = layoutService;
        this.validationService = validationService;
        this.blur = new EventEmitter();
        this.change = new EventEmitter();
        this.customEvent = new EventEmitter();
        this.focus = new EventEmitter();
    }
}
DynamicBootstrapDatePickerComponent.decorators = [
    { type: Component, args: [{
                selector: "dynamic-bootstrap-datepicker",
                template: "<div [formGroup]=\"group\" class=\"input-group\">\n\n    <input bsDatepicker class=\"form-control\" #bsDatepicker=\"bsDatepicker\"\n           [autofocus]=\"model.autoFocus\"\n           [bsConfig]=\"model.getAdditional('bsConfig', {containerClass: model.getAdditional('containerClass', 'theme-green'), showWeekNumbers: model.getAdditional('showWeekNumbers', true)})\"\n           [formControlName]=\"model.id\"\n           [id]=\"elementId\"\n           [maxDate]=\"model.max\"\n           [minDate]=\"model.min\"\n           [name]=\"model.name\"\n           [ngClass]=\"getClass('element', 'control')\"\n           [outsideClick]=\"model.getAdditional('outsideClick', true)\"\n           [placeholder]=\"model.placeholder\"\n           [placement]=\"model.getAdditional('placement', 'bottom')\"\n           [readonly]=\"model.readOnly\"\n           [required]=\"model.required\"\n           (bsValueChange)=\"onChange($event)\"\n           (onHidden)=\"onCustomEvent($event, 'onHidden')\"\n           (onShown)=\"onCustomEvent($event, 'onShown')\">\n\n    <div *ngIf=\"model.toggleLabel\" class=\"input-group-btn\">\n\n        <button *ngIf=\"model.toggleIcon\" class=\"btn\" (click)=\"bsDatepicker.toggle()\"><img [src]=\"model.toggleIcon\"/></button>\n        <button *ngIf=\"model.toggleLabel\" class=\"btn\" (click)=\"bsDatepicker.toggle()\">{{ model.toggleLabel }}</button>\n\n    </div>\n\n</div>\n",
                changeDetection: ChangeDetectionStrategy.OnPush
            }] }
];
/** @nocollapse */
DynamicBootstrapDatePickerComponent.ctorParameters = () => [
    { type: DynamicFormLayoutService },
    { type: DynamicFormValidationService }
];
DynamicBootstrapDatePickerComponent.propDecorators = {
    group: [{ type: Input }],
    layout: [{ type: Input }],
    model: [{ type: Input }],
    blur: [{ type: Output }],
    change: [{ type: Output }],
    customEvent: [{ type: Output }],
    focus: [{ type: Output }],
    bsDatePicker: [{ type: ViewChild, args: [BsDatepickerDirective,] }]
};
if (false) {
    /** @type {?} */
    DynamicBootstrapDatePickerComponent.prototype.group;
    /** @type {?} */
    DynamicBootstrapDatePickerComponent.prototype.layout;
    /** @type {?} */
    DynamicBootstrapDatePickerComponent.prototype.model;
    /** @type {?} */
    DynamicBootstrapDatePickerComponent.prototype.blur;
    /** @type {?} */
    DynamicBootstrapDatePickerComponent.prototype.change;
    /** @type {?} */
    DynamicBootstrapDatePickerComponent.prototype.customEvent;
    /** @type {?} */
    DynamicBootstrapDatePickerComponent.prototype.focus;
    /** @type {?} */
    DynamicBootstrapDatePickerComponent.prototype.bsDatePicker;
    /**
     * @type {?}
     * @protected
     */
    DynamicBootstrapDatePickerComponent.prototype.layoutService;
    /**
     * @type {?}
     * @protected
     */
    DynamicBootstrapDatePickerComponent.prototype.validationService;
}
