// @ng-dynamic-forms/ui-bootstrap/esm2015/src/radio-group/dynamic-bootstrap-radio-group.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output } from "@angular/core";
import { FormGroup } from "@angular/forms";
import { DynamicFormControlComponent, DynamicFormLayoutService, DynamicFormValidationService, DynamicRadioGroupModel } from "@ng-dynamic-forms/core";
export class DynamicBootstrapRadioGroupComponent extends DynamicFormControlComponent {
    /**
     * @param {?} layoutService
     * @param {?} validationService
     */
    constructor(layoutService, validationService) {
        super(layoutService, validationService);
        this.layoutService = layoutService;
        this.validationService = validationService;
        this.blur = new EventEmitter();
        this.change = new EventEmitter();
        this.focus = new EventEmitter();
    }
}
DynamicBootstrapRadioGroupComponent.decorators = [
    { type: Component, args: [{
                selector: "dynamic-bootstrap-radio-group",
                template: "<fieldset role=\"radiogroup\"\n          [formGroup]=\"group\"\n          [id]=\"elementId\"\n          [name]=\"model.name\"\n          [ngClass]=\"getClass('element', 'control')\"\n          [tabindex]=\"model.tabIndex\"\n          (change)=\"onChange($event)\">\n\n    <legend *ngIf=\"model.legend\" [innerHTML]=\"model.legend\"></legend>\n\n    <div *ngFor=\"let option of model.options$ | async\" [ngClass]=\"getClass('element', 'option') || 'radio'\">\n\n        <label><input type=\"radio\"\n                      [formControlName]=\"model.id\"\n                      [name]=\"model.name\"\n                      [value]=\"option.value\"\n                      (blur)=\"onBlur($event)\"\n                      (focus)=\"onFocus($event)\"/><span [innerHTML]=\"option.label\"></span></label>\n    </div>\n\n</fieldset>\n",
                changeDetection: ChangeDetectionStrategy.OnPush
            }] }
];
/** @nocollapse */
DynamicBootstrapRadioGroupComponent.ctorParameters = () => [
    { type: DynamicFormLayoutService },
    { type: DynamicFormValidationService }
];
DynamicBootstrapRadioGroupComponent.propDecorators = {
    group: [{ type: Input }],
    layout: [{ type: Input }],
    model: [{ type: Input }],
    blur: [{ type: Output }],
    change: [{ type: Output }],
    focus: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    DynamicBootstrapRadioGroupComponent.prototype.group;
    /** @type {?} */
    DynamicBootstrapRadioGroupComponent.prototype.layout;
    /** @type {?} */
    DynamicBootstrapRadioGroupComponent.prototype.model;
    /** @type {?} */
    DynamicBootstrapRadioGroupComponent.prototype.blur;
    /** @type {?} */
    DynamicBootstrapRadioGroupComponent.prototype.change;
    /** @type {?} */
    DynamicBootstrapRadioGroupComponent.prototype.focus;
    /**
     * @type {?}
     * @protected
     */
    DynamicBootstrapRadioGroupComponent.prototype.layoutService;
    /**
     * @type {?}
     * @protected
     */
    DynamicBootstrapRadioGroupComponent.prototype.validationService;
}
