// @ng-dynamic-forms/ui-bootstrap/esm2015/src/form-group/dynamic-bootstrap-form-group.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output } from "@angular/core";
import { FormGroup } from "@angular/forms";
import { DynamicFormControlComponent, DynamicFormGroupModel, DynamicFormLayoutService, DynamicFormValidationService } from "@ng-dynamic-forms/core";
export class DynamicBootstrapFormGroupComponent extends DynamicFormControlComponent {
    /**
     * @param {?} layoutService
     * @param {?} validationService
     */
    constructor(layoutService, validationService) {
        super(layoutService, validationService);
        this.layoutService = layoutService;
        this.validationService = validationService;
        this.blur = new EventEmitter();
        this.change = new EventEmitter();
        this.customEvent = new EventEmitter();
        this.focus = new EventEmitter();
    }
}
DynamicBootstrapFormGroupComponent.decorators = [
    { type: Component, args: [{
                selector: "dynamic-bootstrap-form-group",
                template: "<ng-container [formGroup]=\"group\">\n\n    <fieldset [formGroupName]=\"model.id\"\n              [id]=\"elementId\"\n              [name]=\"model.name\"\n              [ngClass]=\"getClass('element', 'control')\">\n\n        <legend *ngIf=\"model.legend\" [innerHTML]=\"model.legend\"></legend>\n\n        <dynamic-bootstrap-form-control *ngFor=\"let _model of model.group\"\n                                        [asBootstrapFormGroup]=\"false\"\n                                        [group]=\"control\"\n                                        [hidden]=\"_model.hidden\"\n                                        [layout]=\"layout\"\n                                        [model]=\"_model\"\n                                        [templates]=\"templates\"\n                                        [ngClass]=\"[getClass('element', 'host', _model), getClass('grid', 'host', _model),\n                                                    getClass('element','children'), getClass('grid','children')]\"\n                                        (blur)=\"onBlur($event)\"\n                                        (change)=\"onChange($event)\"\n                                        (focus)=\"onFocus($event)\"\n                                        (bsEvent)=\"onCustomEvent($event, null, true)\"></dynamic-bootstrap-form-control>\n    </fieldset>\n\n</ng-container>\n"
            }] }
];
/** @nocollapse */
DynamicBootstrapFormGroupComponent.ctorParameters = () => [
    { type: DynamicFormLayoutService },
    { type: DynamicFormValidationService }
];
DynamicBootstrapFormGroupComponent.propDecorators = {
    group: [{ type: Input }],
    layout: [{ type: Input }],
    model: [{ type: Input }],
    templates: [{ type: Input }],
    blur: [{ type: Output }],
    change: [{ type: Output }],
    customEvent: [{ type: Output }],
    focus: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    DynamicBootstrapFormGroupComponent.prototype.group;
    /** @type {?} */
    DynamicBootstrapFormGroupComponent.prototype.layout;
    /** @type {?} */
    DynamicBootstrapFormGroupComponent.prototype.model;
    /** @type {?} */
    DynamicBootstrapFormGroupComponent.prototype.templates;
    /** @type {?} */
    DynamicBootstrapFormGroupComponent.prototype.blur;
    /** @type {?} */
    DynamicBootstrapFormGroupComponent.prototype.change;
    /** @type {?} */
    DynamicBootstrapFormGroupComponent.prototype.customEvent;
    /** @type {?} */
    DynamicBootstrapFormGroupComponent.prototype.focus;
    /**
     * @type {?}
     * @protected
     */
    DynamicBootstrapFormGroupComponent.prototype.layoutService;
    /**
     * @type {?}
     * @protected
     */
    DynamicBootstrapFormGroupComponent.prototype.validationService;
}
