// @ng-dynamic-forms/ui-bootstrap/esm5/src/textarea/dynamic-bootstrap-textarea.component.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output } from "@angular/core";
import { FormGroup } from "@angular/forms";
import { DynamicFormControlComponent, DynamicFormLayoutService, DynamicFormValidationService, DynamicTextAreaModel } from "@ng-dynamic-forms/core";
var DynamicBootstrapTextAreaComponent = /** @class */ (function (_super) {
    __extends(DynamicBootstrapTextAreaComponent, _super);
    function DynamicBootstrapTextAreaComponent(layoutService, validationService) {
        var _this = _super.call(this, layoutService, validationService) || this;
        _this.layoutService = layoutService;
        _this.validationService = validationService;
        _this.blur = new EventEmitter();
        _this.change = new EventEmitter();
        _this.focus = new EventEmitter();
        return _this;
    }
    DynamicBootstrapTextAreaComponent.decorators = [
        { type: Component, args: [{
                    selector: "dynamic-bootstrap-textarea",
                    template: "<ng-container [formGroup]=\"group\">\n\n    <textarea class=\"form-control\"\n              [cols]=\"model.cols\"\n              [formControlName]=\"elementId\"\n              [id]=\"elementId\"\n              [maxlength]=\"model.maxLength\"\n              [minlength]=\"model.minLength\"\n              [name]=\"model.name\"\n              [ngClass]=\"getClass('element', 'control')\"\n              [placeholder]=\"model.placeholder\"\n              [readonly]=\"model.readOnly\"\n              [required]=\"model.required\"\n              [rows]=\"model.rows\"\n              [spellcheck]=\"model.spellCheck\"\n              [tabindex]=\"model.tabIndex\"\n              [wrap]=\"model.wrap\"\n              (blur)=\"onBlur($event)\"\n              (change)=\"onChange($event)\"\n              (focus)=\"onFocus($event)\"></textarea>\n\n</ng-container>\n",
                    changeDetection: ChangeDetectionStrategy.OnPush
                }] }
    ];
    /** @nocollapse */
    DynamicBootstrapTextAreaComponent.ctorParameters = function () { return [
        { type: DynamicFormLayoutService },
        { type: DynamicFormValidationService }
    ]; };
    DynamicBootstrapTextAreaComponent.propDecorators = {
        group: [{ type: Input }],
        layout: [{ type: Input }],
        model: [{ type: Input }],
        blur: [{ type: Output }],
        change: [{ type: Output }],
        focus: [{ type: Output }]
    };
    return DynamicBootstrapTextAreaComponent;
}(DynamicFormControlComponent));
export { DynamicBootstrapTextAreaComponent };
if (false) {
    /** @type {?} */
    DynamicBootstrapTextAreaComponent.prototype.group;
    /** @type {?} */
    DynamicBootstrapTextAreaComponent.prototype.layout;
    /** @type {?} */
    DynamicBootstrapTextAreaComponent.prototype.model;
    /** @type {?} */
    DynamicBootstrapTextAreaComponent.prototype.blur;
    /** @type {?} */
    DynamicBootstrapTextAreaComponent.prototype.change;
    /** @type {?} */
    DynamicBootstrapTextAreaComponent.prototype.focus;
    /**
     * @type {?}
     * @protected
     */
    DynamicBootstrapTextAreaComponent.prototype.layoutService;
    /**
     * @type {?}
     * @protected
     */
    DynamicBootstrapTextAreaComponent.prototype.validationService;
}
