// @ng-dynamic-forms/ui-bootstrap/esm2015/src/checkbox/dynamic-bootstrap-checkbox.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output } from "@angular/core";
import { FormGroup } from "@angular/forms";
import { DynamicCheckboxModel, DynamicFormControlComponent, DynamicFormLayoutService, DynamicFormValidationService, } from "@ng-dynamic-forms/core";
export class DynamicBootstrapCheckboxComponent extends DynamicFormControlComponent {
    /**
     * @param {?} layoutService
     * @param {?} validationService
     */
    constructor(layoutService, validationService) {
        super(layoutService, validationService);
        this.layoutService = layoutService;
        this.validationService = validationService;
        this.blur = new EventEmitter();
        this.change = new EventEmitter();
        this.focus = new EventEmitter();
    }
}
DynamicBootstrapCheckboxComponent.decorators = [
    { type: Component, args: [{
                selector: "dynamic-bootstrap-checkbox",
                template: "<div [formGroup]=\"group\" [class.disabled]=\"model.disabled\" [ngClass]=\"getClass('element', 'control') || 'checkbox'\">\n\n    <label>\n\n        <input type=\"checkbox\"\n               [checked]=\"model.checked\"\n               [formControlName]=\"model.id\"\n               [id]=\"elementId\"\n               [indeterminate]=\"model.indeterminate\"\n               [name]=\"model.name\"\n               [ngClass]=\"getClass('element', 'control')\"\n               [required]=\"model.required\"\n               [tabindex]=\"model.tabIndex\"\n               [value]=\"model.value\"\n               (blur)=\"onBlur($event)\"\n               (change)=\"onChange($event)\"\n               (focus)=\"onFocus($event)\"/><span [innerHTML]=\"model.label\"\n                                                [ngClass]=\"[getClass('element', 'label'), getClass('grid', 'label')]\"></span></label>\n</div>\n",
                changeDetection: ChangeDetectionStrategy.OnPush
            }] }
];
/** @nocollapse */
DynamicBootstrapCheckboxComponent.ctorParameters = () => [
    { type: DynamicFormLayoutService },
    { type: DynamicFormValidationService }
];
DynamicBootstrapCheckboxComponent.propDecorators = {
    group: [{ type: Input }],
    layout: [{ type: Input }],
    model: [{ type: Input }],
    blur: [{ type: Output }],
    change: [{ type: Output }],
    focus: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    DynamicBootstrapCheckboxComponent.prototype.group;
    /** @type {?} */
    DynamicBootstrapCheckboxComponent.prototype.layout;
    /** @type {?} */
    DynamicBootstrapCheckboxComponent.prototype.model;
    /** @type {?} */
    DynamicBootstrapCheckboxComponent.prototype.blur;
    /** @type {?} */
    DynamicBootstrapCheckboxComponent.prototype.change;
    /** @type {?} */
    DynamicBootstrapCheckboxComponent.prototype.focus;
    /**
     * @type {?}
     * @protected
     */
    DynamicBootstrapCheckboxComponent.prototype.layoutService;
    /**
     * @type {?}
     * @protected
     */
    DynamicBootstrapCheckboxComponent.prototype.validationService;
}
